packages/devextreme/js/__internal/grids/grid_core/header_panel/m_header_panel.ts

```diff
@@ -205,33 +205,30 @@ export class HeaderPanel extends ColumnsView {
       args.handled = true;
     }
     if (args.name === 'toolbar') {
-      args.handled = true;
-      if (this._toolbar) {
-        const parts = getPathParts(args.fullName);
-        const optionName = args.fullName.replace(/^toolbar\./, '');
-
-        if (parts.length === 1) {
-          // `toolbar` case
+      const parts = getPathParts(args.fullName);
```

Comment:
It's not required, but let's refactor this code a little to make it more obvious/linear/easy-to-support :)

My suggestion is:
```typescript
  public optionChanged(args: any): void {
    switch (args.name) {
      case 'onToolbarPreparing':
        this._invalidate();
        args.handled = true;
        break;
      case 'toolbar':
        this.optionChangedToolbarHandler(args);
        args.handled = true;
        break;
      default:
        break;
    }

    super.optionChanged(args);
  }

  private optionChangedToolbarHandler(args: any): void {
    const optionNameParts = getPathParts(args.fullName);
    const optionName = args.fullName.replace(/^toolbar\./, '');

    if (optionNameParts[1] === 'items') {
      this.optionChangedToolbarItemsHandler(args.value, optionNameParts, optionName);
      return;
    }

    this._invalidate();
  }

  private optionChangedToolbarItemsHandler(value: unknown, optionNameParts: string[], optionName: string): void {
    // `toolbar.items` case
    if (optionNameParts.length === 2) {
      const { items } = this._getToolbarOptions();
      this._toolbar?.option('items', items);
      return;
    }

    // `toolbar.items[i]` case
    if (optionNameParts.length === 3) {
      const normalizedItem = this._normalizeToolbarItems(this._getToolbarItems(), value);
      this._toolbar?.option(optionName, normalizedItem);
      return;
    }

    // `toolbar.items[i].prop` case
    if (optionNameParts.length >= 4) {
      this._toolbar?.option(optionName, value);
      return;
    }
  }
```

Reply:
Let's do in separate PR later, too much changes for bug fix